cluster/connection_strategy_config.go

```diff
@@ -73,7 +73,8 @@ func (c ConnectionStrategyConfig) Clone() ConnectionStrategyConfig {
 }
 
 func (c *ConnectionStrategyConfig) Validate() error {
-	if err := validate.NonNegativeDuration(&c.Timeout, 1<<63-1, "invalid timeout"); err != nil {
+	// we set the default to 0 in order to be able to override it in failover and cluster configs
+	if err := validate.NonNegativeDuration(&c.Timeout, 0, "invalid timeout"); err != nil {
```

Comment:
As user can set any non-negative value. I think setting it to -1 here makes more sense. 
If user sets it to 0, we can differentiate it. See https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/config/ConnectionRetryConfig.java#L31

Reply:
Why do we allow `0` as a timeout value in the Java client?

Reply:
You can disallow it if you want. then don't check for non-negative but allow only positivies. 
this way, user will not be able to set zero and we can differentiate if it is a user set value or not. 

When it comes to why, it is hard for me to remember the reasoning when we first design it. 
If I try to think something today, a user would set `0` if s/he don't want the client to retry.

Reply:
I see. Since we don't want to get rid of the convenience of public Config, we have to live with it. 
Then, make sure that meaning of zero is well documented. Something like `0 means default and it is infinite for  normal setup and 120 seconds for failover`